cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc

```diff
@@ -0,0 +1,639 @@
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <dirent.h>
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <regex>
+#include <sstream>
+
+#include <encryption-reader-writer-all-crypto-options.h>
+
+/*
+ * This file contains samples for writing and reading encrypted Parquet files in different
+ * encryption and decryption configurations.
+ * Each sample section is dedicated to an independent configuration and shows its creation
+ * from beginning to end.
+ * The samples have the following goals:
+ * 1) Demonstrate usage of different options for data encryption and decryption.
+ * 2) Produce encrypted files for interoperability tests with other (eg parquet-mr)
+ *    readers that support encryption.
+ * 3) Produce encrypted files with plaintext footer, for testing the ability of legacy
+ *    readers to parse the footer and read unencrypted columns.
+ * 4) Perform interoperability tests with other (eg parquet-mr) writers, by reading
+ *    encrypted files produced by these writers.
+ *
+ * Each write sample produces new independent parquet file, encrypted with a different
+ * encryption configuration as described below.
+ * The name of each file is in the form of:
+ * tester<encryption config number>.parquet.encrypted.
+ *
+ * The read sample creates a set of decryption configurations and then uses each of them
+ * to read all encrypted files in the input directory.
+ *
+ * The different encryption and decryption configurations are listed below.
+ *
+ * Usage: ./encryption-interop-tests <write/read> <path-to-directory-of-parquet-files>
+ *
+ * A detailed description of the Parquet Modular Encryption specification can be found
+ * here:
+ * https://github.com/apache/parquet-format/blob/encryption/Encryption.md
+ *
+ * The write sample creates files with four columns in the following
+ * encryption configurations:
+ *
+ *  - Encryption configuration 1:   Encrypt all columns and the footer with the same key.
+ *                                  (uniform encryption)
+ *  - Encryption configuration 2:   Encrypt two columns and the footer, with different
+ *                                  keys.
+ *  - Encryption configuration 3:   Encrypt two columns, with different keys.
```

Comment:
I tried reading this file with the current master and got the following error. Can you take a look?
```
--- Values ---
boolean_field                 |int32_field                   |float_field                   |double_field                  |
1                             |0                             |Parquet error: Couldn't deserialize thrift: don't know what type: 
Deserializing page header failed.

```

Reply:
This branch version says
```
./parquet-reader ./encrypt/tester3.parquet.encrypted 
........ File Metadata ......
Parquet error: Cannot decrypt ColumnMetadata. FileDecryptionProperties must be provided.
```

Reply:
the first error is exactly what should happen when a legacy reader tries to access an encrypted column (it gets a thrift exception on page header).
the second error is from a new reader, trying to read encrypted file without decryption properties.

Reply:
Any reason to enforce decryption properties?
Why cannot we read encrypted parquet files with plaintext footer in legacy mode with the new reader?

Reply:
per the previous comment, we are adding a test option without (with null) FileDecryptionProperties to read encrypted parquet files with plaintext footer in legacy mode with the new reader